### PR TITLE
Add a migrate function to the keeper to clean up state to move to a new gravity contract

### PR DESCRIPTION
**Pull request by zmanian**
This pull request adds supporting infrastructure for cleaning up state and  migrating to new gravity contract.

It does not add the transactions to migrate cosmos ERC20 tokens to the new bridge:

### CODE REVIEW

**Review by mvid — APPROVED** [review]

File: module/x/gravity/keeper/keeper.go [hunk 1]
```diff
@@ -562,3 +562,65 @@ func (k Keeper) CreateContractCallTx(ctx sdk.Context, invalidationNonce uint64,
 	)
 	return newContractCallTx
 }
+
+/////////////////
+// MIGRATE     //
+/////////////////
+
+// Clean up all state associated a previous gravity contract and set a new contract. This is intended to run in the upgrade handler.
+func (k Keeper) MigrateGravityContract(ctx sdk.Context, newBridgeAddress string, bridgeDeploymentHeight uint64) {
```
Comment (jackzampolin) [inline module/x/gravity/keeper/keeper.go:571]: We might want to note that this function doesn't move over any of the ETH based state to the new gravity contract. This is only for the case where there is nothing that needs to be moved over on the ETH side of things. We would likely need separate codes path and flows for:
1. Freezing the gravity contract via governance
2. Requesting the balance of all ERC20 tokens in the gravity contract to be sent via the ETH oracle and verifying with cosmos side supply
3. Outgoing tx to send the ERC20 tokens to the new gravity contract from the old one
That would be in addition to this function which clears state. We might want to add some comments to this effect.

File: module/x/gravity/keeper/keeper.go [hunk 2]
```diff
@@ -562,3 +562,65 @@ func (k Keeper) CreateContractCallTx(ctx sdk.Context, invalidationNonce uint64,
...
+
+	// Delete Any Outgoing TXs.
+
+	prefixStoreOtx := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{types.OutgoingTxKey})
```
Comment (jackzampolin) [inline module/x/gravity/keeper/keeper.go:575]: There is an `k.IterateOutgoingTxs` function too, any reason for not using that here?

File: module/x/gravity/keeper/keeper.go [hunk 3]
```diff
@@ -562,3 +562,65 @@ func (k Keeper) CreateContractCallTx(ctx sdk.Context, invalidationNonce uint64,
...
+
+		for ; iterSig.Valid(); iterSig.Next() {
+			prefixStoreSig.Delete(iterSig.Key())
+		}
+
+		prefixStoreOtx.Delete(iterOtx.Key())
+	}
+
+	//Reset the last observed signer set nonce
+	ctx.KVStore(k.storeKey).Set([]byte{types.LatestSignerSetTxNonceKey}, sdk.Uint64ToBigEndian(0))
```
Comment (jackzampolin) [inline module/x/gravity/keeper/keeper.go:599]: Yes

File: module/x/gravity/keeper/keeper.go (hunk 2 above)
Comment (zmanian) [reply module/x/gravity/keeper/keeper.go:575]: My read of of those iterator functions is that you can't actually delete a value from from the store in those iterators. Is that in correct?

File: module/x/gravity/keeper/keeper.go [hunk 4]
```diff
@@ -562,3 +562,65 @@ func (k Keeper) CreateContractCallTx(ctx sdk.Context, invalidationNonce uint64,
...
+
+	prefixStoreEthereumEvent := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{types.EthereumEventVoteRecordKey})
```
Comment (jackzampolin) [inline module/x/gravity/keeper/keeper.go:601]: k.iterateEthereumEventVoteRecord

Comment (zmanian) [reply module/x/gravity/keeper/keeper.go:601]: I don't think you can delete keys from the iterator.

**Review by jackzampolin — APPROVED** [review]

**Comment by jackzampolin** [comment]
Nice test 💅 ✨

**Pull request merged**